In [1]:
import pandas as pd
import numpy as np
import sklearn 
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [2]:
# Define function to create a binary model from data selected
def fit_binary_model(data,target):
    from sklearn.naive_bayes import BernoulliNB
    bnb = BernoulliNB()
    bnb.fit(data, target)
    pred = bnb.predict(data)
    print('Out of {} predictions, {} were misclassified giving {}% accuracy'.format(data.shape[0], (pred != target).sum(), round((100-((pred != target).sum()/data.shape[0])*100),1)))

In [5]:
# Read in Data
imdb = pd.read_csv('C:\\Users\halln\Desktop\THINKFUL\Datasets\sentiment_detector_model\data_imdb.txt',delimiter='\t',header=None)
amazon = pd.read_csv('C:\\Users\halln\Desktop\THINKFUL\Datasets\sentiment_detector_model\data_amazon.txt',delimiter='\t',header=None)
yelp = pd.read_csv('C:\\Users\halln\Desktop\THINKFUL\Datasets\sentiment_detector_model\data_yelp.txt',delimiter='\t',header=None)

# Change column names to make data uniform
imdb.columns = ['review', 'positive']
yelp.columns = ['review', 'positive']
amazon.columns = ['review', 'positive']

keywords = ['excellent','perfect','terrible', 'awful', 'worst', 'bad', 'stupid', 'poor', 'worse', 'attempt', 'crap', 'fail', 'annoying', 'cheap','painful', 
            'avoid', 'slow', 'pretentious', 'problem', 'embarrassing', 'bored', 'horrible', 'lousy', 'unfortunate', 'boring', 'sucks', 
            'sucked', 'waste',  'wasting', 'mediocre', 'sloppy','disappoint', 'garbage', 'whine','fantastic', 'whiny', 'plot', 'hate ', 'hated',
            'negative', 'nobody', 'flaw', 'script', 'insult', 'torture', ' lack', 'lame', 'ridiculous', 'not', 'unbelievable', 'skip', 'shame', 
           'not even', 'miss', 'amazing', 'love', 'incredible','terrific', 'best', 'great', 'fun',
           'beautiful', 'well done', 'enjoy',  'smart', 'highly', 'impress', 'well',' mess ']

for key in keywords:
    imdb[str(key)] = imdb.review.str.contains(str(key), case = False)
    yelp[str(key)] = yelp.review.str.contains(str(key), case = False)
    amazon[str(key)] = amazon.review.str.contains(str(key), case = False)
    

In [6]:
fit_binary_model(imdb[keywords],imdb['positive'])

Out of 748 predictions, 151 were misclassified giving 79.8% accuracy


In [7]:
fit_binary_model(yelp[keywords],yelp['positive'])

Out of 1000 predictions, 288 were misclassified giving 71.2% accuracy


In [8]:
fit_binary_model(amazon[keywords],amazon['positive'])

Out of 1000 predictions, 273 were misclassified giving 72.7% accuracy


The model works the best for imdb emails with a classification accuracy of 79.8% which makes sense since most of the words in the keywords list were taken from the imdb emails.  The other 2 models still perform fairly well giving >70% accuracy for each